File: src/db/repositories/mcp-config.ts
import { getDb } from "../client.js";
import type { McpConfig, NewMcpConfig } from "../schema.js";

export const mcpConfigRepo = {
  async findByProject(projectId: number): Promise<McpConfig[]> {
    return getDb()
      .selectFrom("mcp_configs")
      .selectAll()
      .where("project_id", "=", projectId)
      .orderBy("server_name")
      .execute();
  },

  async findEnabledByProject(projectId: number): Promise<McpConfig[]> {
    return getDb()
      .selectFrom("mcp_configs")
      .selectAll()
      .where("project_id", "=", projectId)
      .where("enabled", "=", true)
      .execute();
  },

  async upsert(config: NewMcpConfig): Promise<McpConfig> {
    const db = getDb();
    // Try update first
    const existing = await db
      .selectFrom("mcp_configs")
      .selectAll()
      .where("project_id", "=", config.project_id)
      .where("server_name", "=", config.server_name)
      .executeTakeFirst();

    if (existing) {
      return db
        .updateTable("mcp_configs")
        .set({
          config_json: config.config_json,
          enabled: config.enabled,
        })
        .where("id", "=", existing.id)
        .returningAll()
        .executeTakeFirstOrThrow();
    }

    return db
      .insertInto("mcp_configs")
      .values(config)
      .returningAll()
      .executeTakeFirstOrThrow();
  },

  async toggle(id: number, enabled: boolean): Promise<void> {
    await getDb()
      .updateTable("mcp_configs")
      .set({ enabled })
      .where("id", "=", id)
      .execute();
  },

  async deleteById(id: number): Promise<void> {
    await getDb().deleteFrom("mcp_configs").where("id", "=", id).execute();
  },
};

File: proxima/db/repositories/mcp_config.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from sqlalchemy import select, update

from proxima.db.engine import Database
from proxima.db.models import MCPConfig


class MCPConfigRepository:
    def __init__(self, db: Database) -> None:
        self.db = db

    async def find_by_project(self, project_id: int) -> list[MCPConfig]:
        async with self.db.session() as session:
            result = await session.execute(
                select(MCPConfig)
                .where(MCPConfig.project_id == project_id)
                .order_by(MCPConfig.server_name)
            )
            return list(result.scalars().all())

    async def find_enabled_by_project(self, project_id: int) -> list[MCPConfig]:
        async with self.db.session() as session:
            result = await session.execute(
                select(MCPConfig)
                .where(MCPConfig.project_id == project_id)
                .where(MCPConfig.enabled.is_(True))
            )
            return list(result.scalars().all())

    async def upsert(self, values: Mapping[str, Any]) -> MCPConfig:
        async with self.db.session() as session:
            result = await session.execute(
                select(MCPConfig)
                .where(MCPConfig.project_id == int(values["project_id"]))
                .where(MCPConfig.server_name == str(values["server_name"]))
                .limit(1)
            )
            existing = result.scalar_one_or_none()
            if existing:
                existing.config_json = str(values["config_json"])
                existing.enabled = bool(values.get("enabled", True))
                await session.commit()
                await session.refresh(existing)
                return existing

            config = MCPConfig(**dict(values))
            session.add(config)
            await session.commit()
            await session.refresh(config)
            return config

    async def toggle(self, mcp_id: int, enabled: bool) -> None:
        async with self.db.session() as session:
            await session.execute(
                update(MCPConfig).where(MCPConfig.id == mcp_id).values(enabled=enabled)
            )
            await session.commit()

    async def delete_by_id(self, mcp_id: int) -> None:
        async with self.db.session() as session:
            config = await session.get(MCPConfig, mcp_id)
            if config is None:
                return
            await session.delete(config)
            await session.commit()
